docs/design/copy-info.md

```diff
@@ -0,0 +1,271 @@
+below, as well as a new field on Commit objects for backends that support copy
+tracking:
+
+```rust
+/// An individual copy source.
+pub struct CopySource {
+    /// The source path a target was copied from.
+    /// It is not requires that the source path is different than the target
+    /// path. A custom backend may choose to represent 'rollbacks' as copies
+    /// from a file unto itself, from a specific prior commit.
+    path: RepoPathBuf,
+    file: FileId,
+    /// The source commit the target was copied from. If not specified, then the
+    /// parent of the target commit is the source commit. Backends may use this
+    /// field to implement 'integration' logic, where a source may be
+    /// periodically merged into a target, similar to a branch, but the
+    /// branching occurs at the file level rather than the repository level. It
+    /// also follows naturally that any copy source targeted to a specific
+    /// commit should avoid copy propagation on rebasing, which is desirable
+    /// for 'fork' style copies.
+    commit: Option<CommitId>,
+}
+
+using MergedCopySoure = Merge<Option<CopySource>>;
+
+/// An individual copy event, from file A -> B.
+pub struct CopyRecord {
+    /// The destination of the copy, B.
+    target: RepoPathBuf,
+    /// The CommitId where the copy took place.
+    id: CommitId,
+    /// The source of the copy, A.
+    source: MergedCopySoure,
+}
+
+/// Backend options for fetching copy records.
+pub struct CopyRecordOpts {
+    /// If true, follow transitive copy records. That is, if file A is copied to
+    /// B and then to C, a request for copy records of file C should also return
+    /// the copy from A to B. These two copies will be returned in separate
+    /// CopyRecords.
+    transitive: bool
+    // TODO: Probably something for git similarity detection
+}
+
+using CopyRecordStream = Pin<Box<dyn Stream<Item = BackendResult<CopyRecord>>>>;
+
+pub trait Backend {
+    /// Get all copy records for `paths` in the dag range `roots..heads`.
+    /// 
+    /// The exact order these are returned is unspecified, but it is guaranteed
+    /// to be reverse-topological. That is, for any two copy records with
+    /// different commit ids A and B, if A is an ancestor of B, A is streamed
+    /// after B.
+    async fn get_copy_records(&self, paths: &[RepoPathBuf], roots: &[CommitId], heads: &[CommitId]) -> CopyRecordStream;
+}
+```
+
+Obtaining copy records for a single commit requires first computing the files
+list for that commit, then calling get_copy_records with `heads = [id]` and
+`roots = parents()`. This enables commands like `jj diff` to produce better
+diffs that take copy sources into account.
+
+### Write API
+
+Backends that support tracking copy records at the commit level will do so
+through a new field on `backend::Commit` objects:
+
+```rust
+pub struct Commit {
+    ...
+    copies: Option<HashMap<RepoPathBuf, MergedCopySoure>>,
```

Comment:
Hmm, so instead of a Merge<>, we could represent copy info as an ordered set (stored as a Vec<>)?  That's an interesting idea, though some backends (ours definitely) would reject multiple copies in a commit, and I'm not sure how you'd handle that for something like `jj diff` or `jj rebase`.  I chose the Merge<> struct just to represent it as a conflict that needed resolving (where `None` means "no copy source").  Maybe it should still represent that just as a Vec<> instead of a Merge<>, where len()==0 means no copy, len()==1 is resolved and len()>1 is a conflict.

Is your concern with the Merge<> type itself or do you think we should actually support copies from multiple sources in the same commit?

Reply:
> but `CopySource` is a diff at a certain point in the DAG.

Is it, really?
> do you think we should actually support copies from multiple sources in the same commit?

Wouldn't this be like doing `jj new x y`, in which both x and y rename a file to something different, or even F1 to F to F2?

Reply:
This is specifically about two copies _into_ the same destination, since copy info is keyed by destination.  Two separate copies from F -> F1 and F -> F2 can coexist without any complications.

Reply:
OK, so it's like `jj new x y`, in which both x and y rename a file to the same name?

Reply:
> Hmm, so instead of a Merge<>, we could represent copy info as an ordered set (stored as a Vec<>)? That's an interesting idea, though some backends (ours definitely) would reject multiple copies in a commit, and I'm not sure how you'd handle that for something like `jj diff` or `jj rebase`.

If the backend reject multiple sources, `Merge<>` (which is basically `Vec` of positive and negative terms) wouldn't work either. I thought your intention was to support that.

> I chose the Merge<> struct just to represent it as a conflict that needed resolving (where `None` means "no copy source"). Maybe it should still represent that just as a Vec<> instead of a Merge<>, where len()==0 means no copy, len()==1 is resolved and len()>1 is a conflict.
> 
> Is your concern with the Merge<> type itself or do you think we should actually support copies from multiple sources in the same commit?

My concern is what the negative `CopySource` term represents, and where it comes from for the linear squash case. Let's say we have actually squashed a merge commit `A->{B, C}`,
```
         A
        / \
{f<-b} B   C {f<-c}
        \ /
         D {f<-d}
```
we could say that the resulting commit had a copy source `f: Merge[[b, d, c]]` (if we do merge copy sources just like the other value types.) However, I have no idea what the negative term `f: d` would mean. (btw, I think `f: [b, d]` is more correct in this example.)

Reply:
I added a `CopySources` enum to represent resolved/conflicted copies without the negative elements that the `Merge` type has, I think this resolves the main issue.

> If the backend reject multiple sources
To clarify: our backend can support this (and will), but the backend's backend does not and so users will be unable to do the internal equivalent of 'git push' on such commits, which I think is fine (same as you can't push jj commits with conflicts).